datafusion/sql/src/planner.rs

```diff
@@ -236,11 +237,28 @@ impl PlannerContext {
     }
 }
 
+/// This trait allows users to customize the behavior of the SQL planner
```

Comment:
The idea is to pull the custom logic for planning binary operators into a trait -- then we can extend the sql planner by providing trait instances

I think this is beneficial because:
1. It avoids having a DataFusion `Operator` enum that only exists to be transformed later
2. It avoids the planner having hard coded assumptions about function names (like "array_append") 

I think this kind of approach is what might close https://github.com/apache/datafusion/issues/10534

Reply:
I think this seems like a fair approach.

It's worth noting that we could do some of this with #11137 - e.g. we could move the array stuff into a `ParseCustomOperator`, I don't really mind which approach we go with provided the flexibility is there.